File: app/src/main/java/com/sxshi/retrofit/base/HandleException.java
package com.sxshi.retrofit.base;

import org.json.JSONObject;

import java.io.IOException;
import java.net.SocketTimeoutException;

import okhttp3.ResponseBody;
import retrofit2.HttpException;

/**
 * 统一处理异常工具类
 */
public class HandleException {

    private IBaseView baseView;

    public HandleException(IBaseView baseView) {
        this.baseView = baseView;
    }

    public void handlerExcep(Throwable t) {
        if (t instanceof HttpException) {
            ResponseBody responseBody = ((HttpException) t).response().errorBody();
            baseView.onUnknownError(getErrorMessage(responseBody));
        } else if (t instanceof SocketTimeoutException) {
            baseView.onTimeout();
        } else if (t instanceof IOException) {
            baseView.onNetworkError();
        } else {
            baseView.onUnknownError(t.getMessage());
        }
    }

    private String getErrorMessage(ResponseBody responseBody) {
        if (responseBody == null) {
            return "UnknownError";
        }
        try {
            JSONObject jsonObject = new JSONObject(responseBody.string());
            return jsonObject.getString("message");
        } catch (Exception e) {
            return e.getMessage();
        }
    }
}

File: app/src/main/java/com/sxshi/retrofit/base/HandleException.kt
package com.sxshi.retrofit.base

import okhttp3.ResponseBody
import org.json.JSONObject
import retrofit2.HttpException
import java.io.IOException
import java.net.SocketTimeoutException

/**
 * 统一处理异常工具类
 */
class HandleException(private val baseView: IBaseView) {
    fun handlerExcep(t: Throwable) {
        when (t) {
            is HttpException -> {
                val responseBody = t.response()!!.errorBody()
                baseView.onUnknownError(getErrorMessage(responseBody))
            }
            is SocketTimeoutException -> {
                baseView.onTimeout()
            }
            is IOException -> {
                baseView.onNetworkError()
            }
            else -> {
                baseView.onUnknownError(t.message)
            }
        }
    }

    private fun getErrorMessage(responseBody: ResponseBody?): String? {
        return if (responseBody == null) {
            "UnknownError"
        } else try {
            val jsonObject = JSONObject(responseBody.string())
            jsonObject.getString("message")
        } catch (e: Exception) {
            e.message
        }
    }

}